Skip animation update and resolution in subtrees depending on content-visibility

rwlbuis opened the pull request:
#### 37f52f4720baa187a5eb8454ff878840911e6925
<pre>
Skip animation update and resolution in subtrees depending on content-visibility
<a href="https://bugs.webkit.org/show_bug.cgi?id=223930">https://bugs.webkit.org/show_bug.cgi?id=223930</a>

Reviewed by Tim Nguyen.

This change has two parts:
- skip updating and sending events for animation in hidden content-visibility subtrees
- restart the animation once the subtree becomes visible/unskipped

A test is added for the behaviour of content-visibility: auto and animations, one when using scrollIntoView and
a reftest is included testing animations are restarted once a hidden content-visibility subtree becomes visible.

* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/animation-display-lock-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-animation-and-scroll-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-animation-and-scroll.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-animation-becomes-visible-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-animation-becomes-visible.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-animation-in-auto-subtree-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-animation-in-auto-subtree.html: Added.
* Source/WebCore/animation/DocumentTimelinesController.cpp:
(WebCore::DocumentTimelinesController::updateAnimationsAndSendEvents):
* Source/WebCore/animation/WebAnimation.cpp:
(WebCore::WebAnimation::isSkippedContentAnimation const):
* Source/WebCore/animation/WebAnimation.h:
* Source/WebCore/dom/ContentVisibilityDocumentState.cpp:
(WebCore::ContentVisibilityDocumentState::checkRelevancyOfContentVisibilityElement const):
(WebCore::ContentVisibilityDocumentState::skippedContentStateDidChange):
* Source/WebCore/dom/ContentVisibilityDocumentState.h:
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::styleWillChange):

Canonical link: <a href="https://commits.webkit.org/268312@main">https://commits.webkit.org/268312@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/aaae203103efeed469473e338cbb362c02e7ee26

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/19294 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/19714 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/20312 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/21185 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/18053 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/19525 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/22984 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/19837 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/19695 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/19514 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/19565 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/16773 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/22052 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/16751 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/17560 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/23897 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/17813 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/17736 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/21862 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/18327 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/15515 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/17460 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/17397 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4618 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/21820 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/18158 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/rendering/RenderElement.cpp:
```diff
@@ -804,6 +804,7 @@ void RenderElement::styleWillChange(StyleDifference diff, const RenderStyle& new
         if (contentVisibilityChanged) {
             if (oldStyle->contentVisibility() == ContentVisibility::Auto)
                 ContentVisibilityDocumentState::unobserve(*element());
+            ContentVisibilityDocumentState::elementContentVisibilityChange(*element(), oldStyle->contentVisibility(), newStyle.contentVisibility());
```
I'm not super familiar with the animations code in general, but I find it strange that we're doing this based on content visibility style rather than the effective skipped content style. Seems like the spec only mentions skipped content for animations?

Is this something we have test coverage for?

rwlbuis replied:
I am not sure what you mean. The only time skipped content can change is if the root c-v ancestor changes c-v value, which is what is being detected here. Are you thinking of a particular scenario that could fail with this code?

nt1m replied:
@rwlbuis If the c-v root has c-v: auto and changes the content relevancy from relevant to irrelevant, and vice-versa, is that covered? The c-v property isn't changing in that specific case

rwlbuis replied:
You are right, I missed that situation because there was no test. I now added the test and needed  code change.

graouts commented on Source/WebCore/animation/WebAnimation.cpp:
```diff
@@ -1662,4 +1662,13 @@ std::optional<Seconds> WebAnimation::convertAnimationTimeToTimelineTime(Seconds
+        return false;
+    if (auto element = downcast<DeclarativeAnimation>(this)->owningElement())
+        return element->element.renderer() && element->element.renderer()->isSkippedContent();
+    return false;
```
You can use `dynamicDowncast<DeclarativeAnimation>` here.

rwlbuis replied:
Done.

graouts commented on Source/WebCore/dom/ContentVisibilityDocumentState.h:
```diff
@@ -54,6 +54,8 @@ class ContentVisibilityDocumentState {
 
     void updateViewportProximity(const Element&, ViewportProximity);
 
+    static void elementContentVisibilityChange(const Element&, const ContentVisibility&, const ContentVisibility&);
```
I'm not sure whether there is a strict guideline for this, but I feel like we would typically call this method `didChange` to note that the change occurred.

rwlbuis replied:
I like it more, fixed.

graouts commented on Source/WebCore/dom/ContentVisibilityDocumentState.cpp:
```diff
@@ -230,4 +232,23 @@ void ContentVisibilityDocumentState::removeViewportProximity(const Element& elem
+    ASSERT(oldValue != newValue);
+    if (oldValue == newValue || newValue != ContentVisibility::Visible)
+        return;
+    for (auto* animation : WebAnimation::instances()) {
```
If the change happens per document, I feel like this should go through animations registered for this document alone instead of all animation instances. Maybe you could find some use for `Document::matchingAnimations` here?

rwlbuis replied:
The new PR uses matchingAnimations.

rwlbuis replied:
Reverted that as discussed.

graouts commented on Source/WebCore/dom/ContentVisibilityDocumentState.cpp:
```diff
@@ -230,4 +232,20 @@ void ContentVisibilityDocumentState::removeViewportProximity(const Element& elem
+            if (auto* timeline = declarativeAnimation->timeline())
+                timeline->animationTimingDidChange(*declarativeAnimation);
+        }
+    }
```
Looking at the code again, this is not super optimal since we now have two loops. Calling `matchingAnimations` forces a layout which is something important when `Document.getAnimations()` is called via the bindings but not in this case, and in fact could be harmful.

I apologize for suggesting this, upon closer inspection I think your previous code is better. Ideally, we'd find a cleaner way to iterate over animations that both `Document::matchingAnimations()` and this new method would both use, but whatever you had before is satisfactory to me.

rwlbuis replied:
No worries, I reverted the change.

nt1m commented on Source/WebCore/rendering/RenderElement.cpp:
```diff
@@ -813,6 +813,7 @@ void RenderElement::styleWillChange(StyleDifference diff, const RenderStyle& new
         if (contentVisibilityChanged) {
             if (oldStyle->contentVisibility() == ContentVisibility::Auto)
                 ContentVisibilityDocumentState::unobserve(*element());
+            ContentVisibilityDocumentState::skippedContentSubtreeStateDidChange(*element(), oldStyle->contentVisibility() == ContentVisibility::Hidden, newStyle.contentVisibility() == ContentVisibility::Visible);
```
Is it accurate to start skipping immediately if the new value is `auto`? Might be safer / more accurate to not initially skip `auto`

nt1m commented on Source/WebCore/dom/ContentVisibilityDocumentState.cpp:
```diff
@@ -230,4 +233,22 @@ void ContentVisibilityDocumentState::removeViewportProximity(const Element& elem
+
+        auto& declarativeAnimation = downcast<DeclarativeAnimation>(*animation);
+        auto owningElement = declarativeAnimation.owningElement();
+        if (!owningElement || !element.contains(&owningElement->element))
```
The call to `contains()` should use the composed tree / flat tree, since this is rendering code. (Observable differences are notably around slotting / shadow DOM)

nt1m replied:
Maybe @rniwa knows the right DOM method for this.

rwlbuis replied:
I think isDescendantOrShadowDescendantOf does the trick.

nt1m commented on Source/WebCore/dom/ContentVisibilityDocumentState.cpp:
```diff
@@ -230,4 +233,22 @@ void ContentVisibilityDocumentState::removeViewportProximity(const Element& elem
     m_elementViewportProximities.remove(element);
 }
 
+void ContentVisibilityDocumentState::skippedContentSubtreeStateDidChange(const Element& element, bool wasSkipped, bool becomesUnskipped)
```
Is `skippedContentStateDidChange` a better name ? subtree is a bit confusing because this only gets called on the skipped content root nodes.

rwlbuis replied:
I do like it better, adjusted.

nt1m commented on Source/WebCore/animation/WebAnimation.cpp:
```diff
@@ -1662,4 +1662,13 @@ std::optional<Seconds> WebAnimation::convertAnimationTimeToTimelineTime(Seconds
     return animationTime * (1 / m_playbackRate) + *m_startTime;
 }
 
+bool WebAnimation::isInContentVisibilitySubtree() const
```
`isSkippedContentAnimation` would be a more accurate name.

rwlbuis replied:
Sounds good, fixed.

nt1m reviewed (commented):
Can you add a reftest that:
- animation on c-v node
- scroll out of view
- scroll back into view
?

Similarly, can you add a reftest that:
- animation on c-v node
- toggle between hidden/visible
?

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/animation-display-lock-scroll.html:
I was more thinking of reftests to check that the animation resumes (but this test is also fine too)

rwlbuis replied:
To be honest, I am not sure how that would work, like a screenshot but for animations? Do you know of a test example?

nt1m replied:
Something among the lines of using reftest-wait and then pausing the animation, and then screenshotting that.

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-animation-and-scroll.html:
```diff
@@ -0,0 +1,56 @@
+<!DOCTYPE html>
+<html class="reftest-wait">
+<title>Test that making a content-visibility: hidden subtree visible restarts animations in it</title>
```
The title here is inaccurate.

rwlbuis replied:
Good point, fixed.

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-animation-and-scroll.html:
```diff
@@ -0,0 +1,56 @@
+<link rel="author" title="Rob Buis" href="mailto:rbuis@igalia.com">
+<link rel="help" href="https://drafts.csswg.org/css-contain-2/">
+<link rel="match" href="content-visibility-animation-becomes-visible-ref.html">
+<meta name="assert" content="animations in content-visibility: hidden subtree should restart once visible">
```
ditto

rwlbuis replied:
Fixed.

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-animation-and-scroll.html:
```diff
@@ -0,0 +1,56 @@
+}
+.animate {
+  animation: fade 1s linear 2 alternate;
+  animation-delay: 1s;
```
This delay doesn't seem correct to me. You'd screenshot an animation that never started as a result IIUC.

It would be good to have the animation go from transparent to green and then screenshot once the animation ended (e.g. 1s later). That should accurately assert that the animation successfully re-started.

`animation-fill-mode: forwards;` should help taking screenshot a stable state.

nt1m replied:
Would also be good if the animation involved some acceleration in addition to opacity too (transform for instance)

rwlbuis replied:
Fair points, no more delay and I added a transform.

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-animation-becomes-visible.html:
```diff
@@ -0,0 +1,51 @@
+}
+.animate {
+  animation: fade 1s linear 2 alternate;
+  animation-delay: 1s;
```
ditto here.

rwlbuis replied:
Fixed.

nt1m commented on Source/WebCore/dom/ContentVisibilityDocumentState.cpp:
```diff
@@ -140,6 +142,7 @@ bool ContentVisibilityDocumentState::checkRelevancyOfContentVisibilityElement(El
     target.setContentRelevancy(newRelevancy);
     target.invalidateStyle();
     if (target.isConnected()) {
+        skippedContentStateDidChange(target, oldRelevancy->isEmpty(), !newRelevancy.isEmpty());
```
Should this be outside of the `isConnected()` check? I don't really see why it should be inside. Do animations run in disconnected subtrees in general?

rwlbuis replied:
In KeyframeEffect::getKeyframes keyframes for disconnected elements are not included, so in theory this should be safe to do. @graouts WDYT?

nt1m replied:
I don't feel strongly then.

rwlbuis replied:
@graouts we will land as-is but still curious about your answer. If you think it is safe to move it into the "if (isConnected())" section, we can correct that later.

rwlbuis committed:
Skip animation update and resolution in subtrees depending on content-visibility
https://bugs.webkit.org/show_bug.cgi?id=223930

Reviewed by Tim Nguyen.

This change has two parts:
- skip updating and sending events for animation in hidden content-visibility subtrees
- restart the animation once the subtree becomes visible/unskipped

A test is added for the behaviour of content-visibility: auto and animations, one when using scrollIntoView and
a reftest is included testing animations are restarted once a hidden content-visibility subtree becomes visible.

* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/animation-display-lock-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-animation-and-scroll-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-animation-and-scroll.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-animation-becomes-visible-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-animation-becomes-visible.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-animation-in-auto-subtree-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-animation-in-auto-subtree.html: Added.
* Source/WebCore/animation/DocumentTimelinesController.cpp:
(WebCore::DocumentTimelinesController::updateAnimationsAndSendEvents):
* Source/WebCore/animation/WebAnimation.cpp:
(WebCore::WebAnimation::isSkippedContentAnimation const):
* Source/WebCore/animation/WebAnimation.h:
* Source/WebCore/dom/ContentVisibilityDocumentState.cpp:
(WebCore::ContentVisibilityDocumentState::checkRelevancyOfContentVisibilityElement const):
(WebCore::ContentVisibilityDocumentState::skippedContentStateDidChange):
* Source/WebCore/dom/ContentVisibilityDocumentState.h:
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::styleWillChange):

Canonical link: https://commits.webkit.org/268312@main
